### PR TITLE
feat(kuma-cp) extend CircuitBreaker policy with Thresholds

### PR DESCRIPTION
**Pull request by lobkovilya**
### Summary

Today CircuitBreaker supports only Envoy's outlier detector. Current PR adds support of Envoy's circuit breaker thresholds by following the proposal from https://github.com/kumahq/kuma/pull/1655

### Full changelog

* New section `CircuitBreaker#Thresholds`
* Default CircuitBreaker policy
* New Grafana charts:
image

**UPD:**
* Put thresholds on the same "Thresholds Overflow" chart
* Move and rename "Passive Health Check" to "CIrcuit Breakers" row
* use `sum(irate(...))` instead of `increase(...)`
* Duplicate "Circuit Breakers" charts for Mesh dashboard 
image
 
### Documentation

- [X] https://github.com/kumahq/kuma-website/pull/401

### CODE REVIEW

File: pkg/core/resources/apis/mesh/circuit_breaker_validator.go [hunk 1]
```diff
@@ -14,6 +14,13 @@ func (c *CircuitBreakerResource) HasDetectors() bool {
 		c.Spec.Conf.GetDetectors().GetFailure() != nil
 }
 
+func (c *CircuitBreakerResource) HasThresholds() bool {
+	return c.Spec.Conf.GetThresholds().GetMaxConnections() != nil ||
+		c.Spec.Conf.GetThresholds().GetMaxPendingRequests() != nil ||
+		c.Spec.Conf.GetThresholds().GetMaxRetries() != nil ||
+		c.Spec.Conf.GetThresholds().GetMaxRequests() != nil
```
Comment (bartsmykla) [inline pkg/core/resources/apis/mesh/circuit_breaker_validator.go:21]: According to: https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L182-L187

`c.Spec.Conf.GetThresholds()` can return nil

```suggestion
	tresholds := c.Spec.Conf.GetThresholds()

	return tresholds != nil ||
		tresholds.GetMaxConnections() != nil ||
		tresholds.GetMaxPendingRequests() != nil ||
		tresholds.GetMaxRetries() != nil ||
		tresholds.GetMaxRequests() != nil
```

Comment (bartsmykla) [reply pkg/core/resources/apis/mesh/circuit_breaker_validator.go:21]: It would be also good to have test for this I mentioned above

File: pkg/core/resources/apis/mesh/circuit_breaker_validator_test.go [hunk 2]
```diff
@@ -128,6 +134,57 @@ var _ = Describe("CircuitBreaker", func() {
...
+                - match:
+                    kuma.io/service: backend
+                conf:
+                  detectors: {}
+                  thresholds: {}`,
+				expected: `
+               violations:
+               - field: conf
+                 message: must have at least one of the detector or threshold configured`}),
 		)
```
Comment (bartsmykla) [inline pkg/core/resources/apis/mesh/circuit_breaker_validator_test.go:188]: add test when there is no`tresholds`

File: pkg/xds/envoy/clusters/v2/circuit_breaker_configurer.go [hunk 3]
```diff
@@ -3,86 +3,34 @@ package clusters
...
-	} else {
-		outlierDetection.EnforcingFailurePercentage = &wrappers.UInt32Value{Value: 0}
+	cluster.CircuitBreakers = &envoy_cluster.CircuitBreakers{
+		Thresholds: []*envoy_cluster.CircuitBreakers_Thresholds{
+			{
+				Priority:           envoy_api_core.RoutingPriority_DEFAULT,
+				MaxConnections:     c.CircuitBreaker.Spec.Conf.GetThresholds().GetMaxConnections(),
+				MaxPendingRequests: c.CircuitBreaker.Spec.Conf.GetThresholds().GetMaxPendingRequests(),
+				MaxRetries:         c.CircuitBreaker.Spec.Conf.GetThresholds().GetMaxRetries(),
+				MaxRequests:        c.CircuitBreaker.Spec.Conf.GetThresholds().GetMaxRequests(),
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v2/circuit_breaker_configurer.go:31]: 1. What if `c.CircuitBreaker.Spec.Conf.GetThresholds()` would return nil?
2. Maybe we could DRY `c.CircuitBreaker.Spec.Conf.GetThresholds()`

File: pkg/xds/envoy/clusters/v2/outlier_detection_configurer.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	c.configureTotalErrorDetector(cluster.OutlierDetection)
+	c.configureGatewayErrorDetector(cluster.OutlierDetection)
+	c.configureLocalErrorDetector(cluster.OutlierDetection)
+	c.configureStandardDeviationDetector(cluster.OutlierDetection)
+	c.configureFailureDetector(cluster.OutlierDetection)
+	return nil
+}
+
+func (c *OutlierDetectionConfigurer) configureTotalErrorDetector(outlierDetection *envoy_cluster.OutlierDetection) {
+	if total := c.CircuitBreaker.Spec.GetConf().GetDetectors().GetTotalErrors(); total != nil {
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v2/outlier_detection_configurer.go:37]: According to https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L175-L180
`c.CircuitBreaker.Spec.GetConf().GetDetectors()` can return nil

File: pkg/xds/envoy/clusters/v2/outlier_detection_configurer.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+		outlierDetection.Consecutive_5Xx = total.GetConsecutive()
+		outlierDetection.EnforcingConsecutive_5Xx = &wrappers.UInt32Value{Value: 100}
+	} else {
+		outlierDetection.EnforcingConsecutive_5Xx = &wrappers.UInt32Value{Value: 0}
+	}
+}
+
+func (c *OutlierDetectionConfigurer) configureGatewayErrorDetector(outlierDetection *envoy_cluster.OutlierDetection) {
+	if gateway := c.CircuitBreaker.Spec.GetConf().GetDetectors().GetGatewayErrors(); gateway != nil {
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v2/outlier_detection_configurer.go:46]: https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L175-L180

File: pkg/xds/envoy/clusters/v2/outlier_detection_configurer.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+		outlierDetection.ConsecutiveGatewayFailure = gateway.GetConsecutive()
+		outlierDetection.EnforcingConsecutiveGatewayFailure = &wrappers.UInt32Value{Value: 100}
+	} else {
+		outlierDetection.EnforcingConsecutiveGatewayFailure = &wrappers.UInt32Value{Value: 0}
+	}
+}
+
+func (c *OutlierDetectionConfigurer) configureLocalErrorDetector(outlierDetection *envoy_cluster.OutlierDetection) {
+	if local := c.CircuitBreaker.Spec.GetConf().GetDetectors().GetLocalErrors(); local != nil {
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v2/outlier_detection_configurer.go:55]: https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L175-L180

File: pkg/xds/envoy/clusters/v2/outlier_detection_configurer.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		outlierDetection.ConsecutiveLocalOriginFailure = local.GetConsecutive()
+		outlierDetection.EnforcingConsecutiveLocalOriginFailure = &wrappers.UInt32Value{Value: 100}
+	} else {
+		outlierDetection.EnforcingConsecutiveLocalOriginFailure = &wrappers.UInt32Value{Value: 0}
+	}
+}
+
+func (c *OutlierDetectionConfigurer) configureStandardDeviationDetector(outlierDetection *envoy_cluster.OutlierDetection) {
+	if stdev := c.CircuitBreaker.Spec.GetConf().GetDetectors().GetStandardDeviation(); stdev != nil {
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v2/outlier_detection_configurer.go:64]: https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L175-L180

File: pkg/xds/envoy/clusters/v2/outlier_detection_configurer.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+		}
+		outlierDetection.EnforcingSuccessRate = &wrappers.UInt32Value{Value: 100}
+		outlierDetection.EnforcingLocalOriginSuccessRate = &wrappers.UInt32Value{Value: 100} // takes effect only when SplitExternalLocalOriginErrors is true
+	} else {
+		outlierDetection.EnforcingSuccessRate = &wrappers.UInt32Value{Value: 0}
+	}
+}
+
+func (c *OutlierDetectionConfigurer) configureFailureDetector(outlierDetection *envoy_cluster.OutlierDetection) {
+	if failure := c.CircuitBreaker.Spec.GetConf().GetDetectors().GetFailure(); failure != nil {
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v2/outlier_detection_configurer.go:78]: https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L175-L180

File: pkg/xds/envoy/clusters/v2/outlier_detection_configurer_test.go [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+              enforcingFailurePercentageLocalOrigin: 100
+              enforcingFailurePercentage: 100
+              enforcingSuccessRate: 0
+              failurePercentageMinimumHosts: 8
+              failurePercentageRequestVolume: 7
+              failurePercentageThreshold: 85
+            type: EDS`,
+		}),
+	)
+
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v2/outlier_detection_configurer_test.go:251]: Add test case when detectors are `nil`

File: pkg/xds/envoy/clusters/v3/circuit_breaker_configurer.go [hunk 10]
```diff
@@ -2,86 +2,34 @@ package clusters
...
-	} else {
-		outlierDetection.EnforcingFailurePercentage = &wrappers.UInt32Value{Value: 0}
+	cluster.CircuitBreakers = &envoy_cluster.CircuitBreakers{
+		Thresholds: []*envoy_cluster.CircuitBreakers_Thresholds{
+			{
+				Priority:           envoy_config_core_v3.RoutingPriority_DEFAULT,
+				MaxConnections:     c.CircuitBreaker.Spec.Conf.GetThresholds().GetMaxConnections(),
+				MaxPendingRequests: c.CircuitBreaker.Spec.Conf.GetThresholds().GetMaxPendingRequests(),
+				MaxRetries:         c.CircuitBreaker.Spec.Conf.GetThresholds().GetMaxRetries(),
+				MaxRequests:        c.CircuitBreaker.Spec.Conf.GetThresholds().GetMaxRequests(),
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v3/circuit_breaker_configurer.go:30]: According to: https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L182-L187
`c.CircuitBreaker.Spec.Conf.GetThresholds()` can be nil

Also, maybe we could DRY it?

File: pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+	c.configureTotalErrorDetector(cluster.OutlierDetection)
+	c.configureGatewayErrorDetector(cluster.OutlierDetection)
+	c.configureLocalErrorDetector(cluster.OutlierDetection)
+	c.configureStandardDeviationDetector(cluster.OutlierDetection)
+	c.configureFailureDetector(cluster.OutlierDetection)
+	return nil
+}
+
+func (c *OutlierDetectionConfigurer) configureTotalErrorDetector(outlierDetection *envoy_cluster.OutlierDetection) {
+	if total := c.CircuitBreaker.Spec.GetConf().GetDetectors().GetTotalErrors(); total != nil {
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go:36]: According to https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L175-L180
`c.CircuitBreaker.Spec.GetConf().GetDetectors()` can return nil

File: pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+		outlierDetection.Consecutive_5Xx = total.GetConsecutive()
+		outlierDetection.EnforcingConsecutive_5Xx = &wrappers.UInt32Value{Value: 100}
+	} else {
+		outlierDetection.EnforcingConsecutive_5Xx = &wrappers.UInt32Value{Value: 0}
+	}
+}
+
+func (c *OutlierDetectionConfigurer) configureGatewayErrorDetector(outlierDetection *envoy_cluster.OutlierDetection) {
+	if gateway := c.CircuitBreaker.Spec.GetConf().GetDetectors().GetGatewayErrors(); gateway != nil {
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go:45]: According to https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L175-L180
`c.CircuitBreaker.Spec.GetConf().GetDetectors()` can return nil

File: pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+		outlierDetection.ConsecutiveGatewayFailure = gateway.GetConsecutive()
+		outlierDetection.EnforcingConsecutiveGatewayFailure = &wrappers.UInt32Value{Value: 100}
+	} else {
+		outlierDetection.EnforcingConsecutiveGatewayFailure = &wrappers.UInt32Value{Value: 0}
+	}
+}
+
+func (c *OutlierDetectionConfigurer) configureLocalErrorDetector(outlierDetection *envoy_cluster.OutlierDetection) {
+	if local := c.CircuitBreaker.Spec.GetConf().GetDetectors().GetLocalErrors(); local != nil {
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go:54]: According to https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L175-L180
`c.CircuitBreaker.Spec.GetConf().GetDetectors()` can return nil

File: pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+		outlierDetection.ConsecutiveLocalOriginFailure = local.GetConsecutive()
+		outlierDetection.EnforcingConsecutiveLocalOriginFailure = &wrappers.UInt32Value{Value: 100}
+	} else {
+		outlierDetection.EnforcingConsecutiveLocalOriginFailure = &wrappers.UInt32Value{Value: 0}
+	}
+}
+
+func (c *OutlierDetectionConfigurer) configureStandardDeviationDetector(outlierDetection *envoy_cluster.OutlierDetection) {
+	if stdev := c.CircuitBreaker.Spec.GetConf().GetDetectors().GetStandardDeviation(); stdev != nil {
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go:63]: According to https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L175-L180
`c.CircuitBreaker.Spec.GetConf().GetDetectors()` can return nil

File: pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+		}
+		outlierDetection.EnforcingSuccessRate = &wrappers.UInt32Value{Value: 100}
+		outlierDetection.EnforcingLocalOriginSuccessRate = &wrappers.UInt32Value{Value: 100} // takes effect only when SplitExternalLocalOriginErrors is true
+	} else {
+		outlierDetection.EnforcingSuccessRate = &wrappers.UInt32Value{Value: 0}
+	}
+}
+
+func (c *OutlierDetectionConfigurer) configureFailureDetector(outlierDetection *envoy_cluster.OutlierDetection) {
+	if failure := c.CircuitBreaker.Spec.GetConf().GetDetectors().GetFailure(); failure != nil {
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v3/outlier_detection_configurer.go:77]: According to https://github.com/kumahq/kuma/blob/ff2e4f2eab0d5da3f08397692a4c69c2b9a9513d/api/mesh/v1alpha1/circuit_breaker.pb.go#L175-L180
`c.CircuitBreaker.Spec.GetConf().GetDetectors()` can return nil

File: pkg/xds/envoy/clusters/v3/outlier_detection_configurer_test.go [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+              enforcingFailurePercentageLocalOrigin: 100
+              enforcingFailurePercentage: 100
+              enforcingSuccessRate: 0
+              failurePercentageMinimumHosts: 8
+              failurePercentageRequestVolume: 7
+              failurePercentageThreshold: 85
+            type: EDS`,
+		}),
+	)
+})
```
Comment (bartsmykla) [inline pkg/xds/envoy/clusters/v3/outlier_detection_configurer_test.go:259]: Add test case when detectors are `nil`

**Review by bartsmykla — CHANGES_REQUESTED** [review]

File: pkg/core/resources/apis/mesh/circuit_breaker_validator.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/core/resources/apis/mesh/circuit_breaker_validator.go:21]: That's true, but `GetMaxConnections` will return `nil` if `thresholds` is `nil`. Same for all other autogenerated proto methods. I think we can rely on it and have fewer nil-checks when it comes to protubuf.

File: pkg/core/resources/apis/mesh/circuit_breaker_validator_test.go (hunk 2 above)
Comment (lobkovilya) [reply pkg/core/resources/apis/mesh/circuit_breaker_validator_test.go:199]: test `one detector with default values` already checks it

File: pkg/xds/envoy/clusters/v2/circuit_breaker_configurer.go (hunk 3 above)
Comment (lobkovilya) [reply pkg/xds/envoy/clusters/v2/circuit_breaker_configurer.go]: 1. same, I don't think we need nil-checks here
2. I don't mind to have a variable `thresholds`, fixed

File: pkg/xds/envoy/clusters/v2/outlier_detection_configurer_test.go (hunk 9 above)
Comment (lobkovilya) [reply pkg/xds/envoy/clusters/v2/outlier_detection_configurer_test.go:251]: Added test `only thresholds`

File: pkg/core/resources/apis/mesh/circuit_breaker_validator_test.go (hunk 2 above)
Comment (bartsmykla) [reply pkg/core/resources/apis/mesh/circuit_breaker_validator_test.go:199]: 👍

File: pkg/core/resources/apis/mesh/circuit_breaker_validator.go (hunk 1 above)
Comment (bartsmykla) [reply pkg/core/resources/apis/mesh/circuit_breaker_validator.go:21]: You are right. I realized it's checking for `x` and not `x.Thresholds`.

**Review by bartsmykla — APPROVED** [review]

File: pkg/defaults/mesh/circuit_breaker.go [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+		return nil, false
+	}
+	if !store.IsResourceNotFound(err) {
+		return errors.Wrap(err, "could not retrieve a resource"), false
+	}
+	if err := resManager.Create(context.Background(), circuitBreaker, store.CreateBy(key)); err != nil {
+		return errors.Wrap(err, "could not create a resource"), false
+	}
+	return nil, true
+}
```
Comment (jakubdyszkiewicz) [inline pkg/defaults/mesh/circuit_breaker.go:58]: should this go into the release branch?

**Review by jakubdyszkiewicz — APPROVED** [review]
